go/vt/mysqlctl/compression.go

```diff
@@ -0,0 +1,310 @@
+}
+
+// This is a wrapper to get the right decompressor (see below) based on the extension of the file.
+func newBuiltinDecompressorFromExtension(extension, engine string, reader io.Reader, logger logutil.Logger) (decompressor io.ReadCloser, err error) {
+	// we only infer the engine from the extension is set to "auto", otherwise we use whatever the user selected
+	if engine == "auto" {
+		logger.Infof("Builtin decompressor set to auto, checking which engine to decompress based on the extension")
+
+		eng, err := getEngineFromExtension(extension)
+		if err != nil {
+			return decompressor, err
+		}
+
+		engine = eng
+	}
+
+	return newBuiltinDecompressor(engine, reader, logger)
+}
+
+// This returns a reader that will decompress the underlying provided reader and will use the specified supported engine.
+func newBuiltinDecompressor(engine string, reader io.Reader, logger logutil.Logger) (decompressor io.ReadCloser, err error) {
+	if engine == "pargzip" {
+		logger.Warningf("engine \"pargzip\" doesn't support decompression, using \"pgzip\" instead")
+
+		engine = "pgzip"
+	}
+
+	switch engine {
+	case "pgzip":
+		d, err := pgzip.NewReader(reader)
+		if err != nil {
+			return nil, err
+		}
+
+		decompressor = d
+	case "lz4":
+		decompressor = ioutil.NopCloser(lz4.NewReader(reader))
+	case "zstd":
+		d, err := zstd.NewReader(reader)
+		if err != nil {
+			return nil, err
+		}
+
+		decompressor = d.IOReadCloser()
+	default:
+		err = fmt.Errorf("Unkown decompressor engine: %q", engine)
+		return decompressor, err
+	}
+
+	logger.Infof("Decompressing backup using engine %q", engine)
+
+	return decompressor, err
+}
+
+// This returns a writer that will compress the data using the specified engine before writing to the underlying writer.
+func newBuiltinCompressor(engine string, writer io.Writer, logger logutil.Logger) (compressor io.WriteCloser, err error) {
+	switch engine {
+	case "pgzip":
+		gzip, err := pgzip.NewWriterLevel(writer, *compressionLevel)
+		if err != nil {
+			return compressor, vterrors.Wrap(err, "cannot create gzip compressor")
+		}
+
+		gzip.SetConcurrency(*backupCompressBlockSize, *backupCompressBlocks)
+		compressor = gzip
+	case "pargzip":
+		gzip := pargzip.NewWriter(writer)
+		gzip.ChunkSize = *backupCompressBlockSize
+		gzip.Parallel = *backupCompressBlocks
+		gzip.CompressionLevel = *compressionLevel
+		compressor = gzip
+	case "lz4":
+		lz4Writer := lz4.NewWriter(writer).WithConcurrency(*backupCompressBlocks)
+		lz4Writer.Header = lz4.Header{
+			CompressionLevel: *compressionLevel,
+		}
+		compressor = lz4Writer
+	case "zstd":
+		zst, err := zstd.NewWriter(writer, zstd.WithEncoderLevel(zstd.EncoderLevel(*compressionLevel)))
+		if err != nil {
+			return compressor, vterrors.Wrap(err, "cannot create zstd compressor")
+		}
+
+		compressor = zst
+	default:
+		err = fmt.Errorf("Unkown compressor engine: %q", engine)
+		return compressor, err
+	}
+
+	logger.Infof("Compressing backup using engine %q", engine)
+
```

Comment:
```suggestion
```

Reply:
I realize this is more stylistic than anything else (and I have no idea what the official golfing style guide would say), but longer files / funcs are just harder to read on a screen where you only see a limited number of lines at a time.

Reply:
I think a lot of those might be here because of our linter configuration for `golangci-lint`, so definitely remove them if it fits better the vitess codebase.